enhancements/authentication/pod-security-admission-autolabeling.md

```diff
@@ -0,0 +1,290 @@
+basis by using the `pod-security.kubernetes.io/enforce` label on given namespaces. It is also
+possible to exampt certain users, namespace and runtime classes from the admission completely.
+
+#### PSa Label Synchronization Controller
+
+This section describes a controller that is going to be a part of the cluster-policy-controller.
+
+In order to keep SCCs working alongside restricted cluster-wide PSa level, there must be
+a mechanism that would synchronize the PSa policy level for a namespace that workload runs
+in so that the workload is allowed by both SCC and PSa admissions. For that purpose, a PSa
+Label Synchronization Controller (hereinafter the Controller) shall be built.
+
+The Controller watches changes of:
+- **namespaces**
+  - manual changes to the interesting labels of the namespaces that were
+    opted-in to label synchronization should be reverted
+- all known **workload types** in controlled namespaces
+  - new pods may be getting rejected by PSa for some time before the label
+    synchronization happens
+  - the controller should not allow more relaxed PSa admission level than
+    what SCCs allow
+- **roles** and **rolebindings** (and their **cluster variants**)
+  - to be able to properly assess SA capabilities in controlled namespaces
+- **SCCs**
+  - to properly assess restrictive PSa levels per affected namespaces
+
+The generic control loop of the Controller performs the following actions:
+1. list all controlled (opted-in by a label) namespaces
+2. for each namespace:
+    1. get all instances of known workload types
+    1. associate SCCs to the workloads based on the SAs running them, create a
+       SCC <-> []SA associations
+    1. order observed SCCs most-privileged first (according to PSa privilege levels
+       they map to)
+    1. evaluate the namespace's PSa privilege level based on the most privileged
+       SCC that is allowed to be used by its SA
+    1. modify the namespace's PSa labels based on the enforce level obsedved in the
+       previous step
+
+There is another control loop reacting only to changes of namespaces which performs
+all the steps in 2.
+
+#### Label Synchronization Opting-in
+
+The Controller described in [PSa Label Synchronization Controller](#psa-label-synchronization-controller)
+works on an opt-in basis, meaning that users need to specifically sign themselves up
+if they wanted to stick with just SCCs.
+
+The opt-in is driven by the `security.openshift.io/scc.podSecurityLabelSync: “true”`
+label, namespaces containing it will have their PSa labels reconciled by
+the Controller.
+
+The opt-in nature of the Controller would cause current workloads to break.
+The choice of opting-in is therefore only left upon the users that are
+capable of creating the `Namespace` objects directly. Such a user can clearly
+set the appropriate labels themselves.
+
+The rest of the users rely on `Namespace` creation delegated via the OpenShift
```

Comment:
This is confusing, first you're talking about the mechanism being opt-in, but eventually this is only opt-in to a small, privileged group of users, since you're changing the default. In that case I'd start with this part where you say we make it a default, but only a set of users can turn it off. Or am I misunderstanding something here 😅

Reply:
Hm, you're right, I'll have to reword this section somewhat more understandable.

The truth indeed is that the feature is opt-in only for the privileged users, the rest gets the "old" behavior. And understandably so, we don't want just anyone be choosing which admission to use for their pods, that's a consideration for a more privileged user.